### DISCUSSION TITLE
5 seconds delay for new/modify entry PUSH NOTIFICATION

### DISCUSSION THREAD

**Discussion by kamilos956** [dani-garcia/vaultwarden#4523] [Ideas]
### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.30.0
* Web-vault version: v2023.10.0
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.41.2
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 

```json
{
  "_duo_akey": null,
  "_enable_duo": true,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://***************",
  "domain_origin": "*****://***************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_change_allowed": true,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": false,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "debug",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": true,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": false,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": true,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": "Plain",
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "**********************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "**************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "**********************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

### Steps to reproduce
I used application 1-2 days, login first to MacBook and iPhone, then login to Brave. All is working fine, new / modify entry saving immediately. After 1 day of use, I have 5 seconds delay to edit / new entry saving. No matter is phone or browser. While delay saving icon freeze application and after that, works fine. What I found, I have this delay only while PUSH NOTIFICATION are enabled, when I disable all always is working fine. For this delay helps deauthorized all sessions from admin panel. After that is ok for another 1-2 days.

### Expected behaviour
Always working with no delay. 

### Actual behaviour
5 seconds delay for new / modify entry. As you see below, ~4 seconds pause after created socket succesfully.

### Troubleshooting data
[2023-11-18 08:49:43.387][request][INFO] DELETE /api/ciphers/07c621ce-d875-414a-8bdd-bf79f5c5b05b
[2023-11-18 08:49:43.436][reqwest::connect][DEBUG] starting new connection: https://identity.bitwarden.com/
[2023-11-18 08:49:43.436][trust_dns_proto::xfer::dns_handle][DEBUG] querying: identity.bitwarden.com A
[2023-11-18 08:49:43.436][trust_dns_resolver::name_server::name_server_pool][DEBUG] sending request: [Query { name: Name("identity.bitwarden.com"), query_type: A, query_class: IN }]
[2023-11-18 08:49:43.436][trust_dns_resolver::name_server::name_server][DEBUG] reconnecting: NameServerConfig { socket_addr: 127.0.0.11:53, protocol: Udp, tls_dns_name: None, trust_negative_responses: false, bind_addr: None }
[2023-11-18 08:49:43.436][trust_dns_proto::xfer][DEBUG] enqueueing message:QUERY:[Query { name: Name("identity.bitwarden.com"), query_type: A, query_class: IN }]
[2023-11-18 08:49:43.436][trust_dns_proto::udp::udp_client_stream][DEBUG] final message: ; header 2751:QUERY:RD:NoError:QUERY:0/0/0
; query
;; identity.bitwarden.com. IN A

[2023-11-18 08:49:43.436][trust_dns_proto::udp::udp_stream][DEBUG] created socket successfully
[2023-11-18 08:49:47.453][trust_dns_proto::udp::udp_client_stream][DEBUG] received message id: 2751
[2023-11-18 08:49:47.454][trust_dns_resolver::error][DEBUG] Response:; header 2751:RESPONSE:RD,RA:NoError:QUERY:2/0/0
; query
;; identity.bitwarden.com. IN A
; answers 2

**Comment by nomad2246** [comment]
Hi,
I had this same problem, or a very similar one.
I had the DNS of the container itself pointing to the server where it is hosted, the latter picked up the requests and sent them to cloudflared encrypted (DoH).
As soon as I point the DNS of the container directly to cloudflared (1.1.1.1.1/1.0.0.1) the delay when saving/modifying a vault element is over.

**Comment by BlackDex** [comment]
This could indeed be DNS related. I'm not really able to mimic the exact same though.

I'm going to move this too the Idea section. The only thing what might solve this specific item is that we somehow decouple the notifications code via a thread or something from the rest of the web-vault.

Since this is more of a performance/UX item and not a real bug ill remove it from issues.
